src/TestFramework/Coverage/JUnit/JUnitReportLocator.php

```diff
@@ -0,0 +1,122 @@
+ *
+ * Copyright (c) 2017, Maks Rafalko
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * * Redistributions of source code must retain the above copyright notice, this
+ *   list of conditions and the following disclaimer.
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+declare(strict_types=1);
+
+namespace Infection\TestFramework\Coverage\JUnit;
+
+use function array_map;
+use function count;
+use function current;
+use function file_exists;
+use function implode;
+use Infection\FileSystem\Locator\FileNotFound;
+use function iterator_to_array;
+use function Safe\sprintf;
+use Symfony\Component\Finder\Finder;
+use Symfony\Component\Finder\SplFileInfo;
+use Webmozart\PathUtil\Path;
+
+/**
+ * @internal
+ * @final
+ */
+class JUnitReportLocator
+{
+    private $coveragePath;
+    private $defaultJUnitPath;
+
+    /**
+     * @var string|null
+     */
+    private $jUnitPath;
+
+    public function __construct(string $coveragePath, string $defaultJUnitPath)
+    {
+        $this->coveragePath = Path::canonicalize($coveragePath . '/..');
+        $this->defaultJUnitPath = Path::canonicalize($defaultJUnitPath);
+    }
+
+    /**
+     * @throws FileNotFound
+     */
+    public function locate(): string
+    {
+        if ($this->jUnitPath !== null) {
+            return $this->jUnitPath;
+        }
+
+        // This is the JUnit path enforced before. It is also the one recommended by the
+        // CoverageChecker hence it makes sense to try this one first before attempting any more
+        // expensive lookup
+        if (file_exists($this->defaultJUnitPath)) {
+            return $this->jUnitPath ?? $this->defaultJUnitPath;
```

Comment:
do we need to check `$this->jUnitPath` here? If it's not `null`, we will early return on line 76 I guess

Reply:
hm... `$this->jUnitPath` is never assigned, probably you meant to use `??=` here instead of `??`: https://3v4l.org/gpRX9

Then the test that checks memozing does not work correctly, I guess with current state of this PR nothing is cached

Reply:
>Then the test that checks memozing does not work correctly, I guess with current state of this PR nothing is cached

Might not be limited to here... Will double check that elsewhere as well in another PR

Reply:
Nvm just need to be `=` here